### PR TITLE
Fix importing units from a common base model

### LINKED ISSUE #1160: Importing Units from a common resource
When using a model as a database for units, the units do not get resolved as expected.
For example, given a model:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#" name="triangle_point_I">
  <import xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="triangle_units_base.cellml">
    <units units_ref="mm" name="mm"/>
  </import>
  <import xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="triangle_units_opposite_I.cellml">
    <component component_ref="opposite" name="opposite"/>
  </import>
  <component name="point">
    <variable name="var2" units="mm" />
  </component>
</model>
```
and the imported units model:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#" name="diamond_units_base">
  <units name="mm">
    <unit prefix="milli" units="metre"/>
  </units>
</model>
```
and the imported component model:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#" name="triangle_opposite">
  <import xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="triangle_units_base.cellml">
    <units units_ref="mm" name="mm"/>
  </import>
  <component name="opposite">
    <variable name="var1" units="mm" />
  </component>
</model>
```

Will result in the following model:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#" name="triangle_point_I">
  <units name="mm">
    <unit prefix="milli" units="metre"/>
  </units>
  <units name="mm_1">
    <unit prefix="milli" units="metre"/>
  </units>
  <component name="opposite">
    <variable name="var1" units="mm_1"/>
  </component>
  <component name="point">
    <variable name="var2" units="mm"/>
  </component>
</model>
```
Which is not the desired outcome.
The desired output would be:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#" name="triangle_point_I">
  <units name="mm">
    <unit prefix="milli" units="metre"/>
  </units>
  <component name="opposite">
    <variable name="var1" units="mm"/>
  </component>
  <component name="point">
    <variable name="var2" units="mm"/>
  </component>
</model>
```

### PR DESCRIPTION
**Pull request by hsorby**
Fixes #1160.

### CODE REVIEW

**Comment by hsorby** [comment]
Based on #1157, so that needs attention to make this PR review simpler.

File: src/importer.cpp [hunk 1]
```diff
@@ -689,6 +689,28 @@ std::string addUnitsAvoidingNameClash(const ModelPtr &model, const UnitsPtr &uni
...
+{
+    if (u->isImport()) {
+        auto importSource = u->importSource();
+        auto importingModel = importSource->model();
+        auto importedUnits = importingModel->units(u->importReference());
+        auto importedUnitsCopy = importedUnits->clone();
+        importedUnitsCopy->setName(u->name());
+        model->replaceUnits(index, importedUnitsCopy);
+        for (size_t unitIndex = 0; unitIndex < importedUnits->unitCount(); ++unitIndex) {
+            const std::string reference = importedUnits->unitAttributeReference(unitIndex);
```
Comment (agarny) [inline src/importer.cpp:702]: Might want to replace `const std::string` with `auto`.

File: src/importer.cpp [hunk 2]
```diff
@@ -689,6 +689,28 @@ std::string addUnitsAvoidingNameClash(const ModelPtr &model, const UnitsPtr &uni
...
+            if (!reference.empty() && !isStandardUnitName(reference)) {
+                if (importingModel->hasUnits(reference)) {
```
Comment (agarny) [inline src/importer.cpp]: I would combine those two `if` statements.

File: src/importer.cpp [hunk 3]
```diff
@@ -767,11 +789,29 @@ ComponentPtr flattenComponent(const ComponentEntityPtr &parent, ComponentPtr &co
...
+                    flattenUnitsTree(clonedImportModel, u, unitsIndex);
+                    flattenedUnits = clonedImportModel->units(unitsIndex);
+                    break;
+                }
+            }
+
+            auto replacementUnits = (flattenedUnits != nullptr) ? flattenedUnits : u;
+
+            const std::string originalName = replacementUnits->name();
+            const std::string newName = addUnitsAvoidingNameClash(model, replacementUnits);
```
Comment (agarny) [inline src/importer.cpp:810]: Would remove the above empty line and replace `const std::string` with `auto`.

**Review by nickerso — DISMISSED** [review]

File: src/validator.cpp [hunk 4]
```diff
@@ -683,7 +684,7 @@ class Validator::ValidatorImpl: public LoggerImpl
 bool checkForLocalCycles(const History &history, const HistoryEpochPtr &h)
 {
     return std::find_if(history.begin(), history.end(),
-                        [=](const HistoryEpochPtr &i) -> bool { return i->mName == h->mName; })
+                        [=](const HistoryEpochPtr &i) -> bool { return i->mName == h->mName && i->mSourceUrl == h->mSourceUrl; })
```
Comment (agarny) [inline src/validator.cpp]: `return (i->mName == h->mName) && (i->mSourceUrl == h->mSourceUrl);`

**Review by agarny — CHANGES_REQUESTED** [review]

File: src/validator.cpp (hunk 4 above)
Comment (hsorby) [reply src/validator.cpp]: [b8934be](https://github.com/cellml/libcellml/pull/1161/commits/b8934bea3a47d12c2648865353dd4c240e2395d4)

File: src/importer.cpp (hunk 2 above)
Comment (hsorby) [reply src/importer.cpp]: [991b722](https://github.com/cellml/libcellml/pull/1161/commits/991b7227bba155a34dd476bb9ce587f035c97e40)

File: src/importer.cpp (hunk 1 above)
Comment (hsorby) [reply src/importer.cpp:702]: I quite like my `const std::string`.

File: src/importer.cpp (hunk 3 above)
Comment (hsorby) [reply src/importer.cpp:810]: I have the empty line so that the above statement is isolated and clear.

**Review by agarny — APPROVED** [review]

**Review by nickerso — APPROVED** [review]

**Pull request merged**